src/translation.cpp

```diff
@@ -0,0 +1,760 @@
+				lcf::rpg::EventCommand newCmd;
+				newCmd.code = static_cast<int>(baseMsgBox ? lcf::rpg::EventCommand::Code::ShowMessage : lcf::rpg::EventCommand::Code::ShowMessage_2);
+				newCmd.indent = refIndent;
+				newCmd.string = lcf::DBString(line);
+				commands.insert(commands.begin()+idx, newCmd);
+
+				// Update our index
+				if (index >= idx) {
+					index += 1;
+				}
+			}
+		}
+
+		/// Remove the EventCommand at position "idx" from the command stack.
+		/// Also updates our index, if relevant.
+		void RemoveByIndex(size_t idx) {
+			if (idx < commands.size()) {
+				commands.erase(commands.begin() + idx);
+
+				// Update our index
+				if (index > idx) {
+					index -= 1;
+				}
+			}
+		}
+
+		/// Add multiple message boxes to the command stack before "idx".
+		/// The "msgs" each represent lines in new, independent message boxes (so they will have both ShowMessage and ShowMessage_2)
+		/// Also updates our index, if relevant.
+		void InsertMultiMessageBefore(std::vector<std::vector<std::string>>& msgs, size_t idx) {
+			for (std::vector<std::string>& lines : msgs) {
+				while (!lines.empty()) {
+					PutShowMessageBeforeIndex(lines.back(), idx, lines.size()==1);
+					lines.pop_back();
+				}
+			}
+		}
+
+	private:
+		std::vector<lcf::rpg::EventCommand>& commands;
+		size_t index = 0;
+	};
+}
+
+
+
+std::vector<std::vector<std::string>> Translation::TranslateMessageStream(const Dictionary& dict, const std::stringstream& msg, char trimChar) {
+	// Prepare source string
+	std::string msgStr = msg.str();
+	if (msgStr.size()>0 && msgStr.back() == trimChar) {
+		msgStr.pop_back();
+	}
+
+	// Translation exists?
+	std::vector<std::vector<std::string>> res;
+	if (dict.TranslateString<std::string>("", msgStr)) {
+		// First, get all lines.
+		std::vector<std::string> lines = Utils::Tokenize(msgStr, [](char32_t c) { return c=='\n'; });
+
+		// Now, break into message boxes based on the ADDMSG string
+		res.push_back(std::vector<std::string>());
+		for (const std::string& line : lines) {
+			if (line == TRCUST_ADDMSG) {
+				res.push_back(std::vector<std::string>());
+			} else {
+				res.back().push_back(line);
+			}
+
+			// Special case: stop once you've found a REMMSG (to avoid the case where both add/rem are present)
+			if (line == TRCUST_REMOVEMSG) {
+				break;
+			}
+		} 
+
+		// Ensure we never get an empty vector (force using the REMMSG command)
+		for (std::vector<std::string>& msgbox : res) {
+			if (msgbox.empty()) {
+				msgbox.push_back("");
+			}
+		}
+	}
+
+	return res;
+}
+
+
+void Translation::RewriteEventCommandMessage(const Dictionary& dict, std::vector<lcf::rpg::EventCommand>& commandsOrig) {
+	// A note on this function: it is (I feel) necessarily complicated, given what it's actually doing.
+	// I've tried to abstract most of this complexity away behind an "iterator" interface, so that we do not
+	// have to track the current index directly in this function.
+	CommandIterator commands(commandsOrig);
```

Comment:
Marking this as "Reviewing later", can you explain me first how this TRCUST_REMOVEMSG stuff works? Considering that this is never pushed anywhere I'm confused.

Reply:
See the earlier comment about the translator adding the strings into the translated text. 

Given that, the basic idea is to take a CommandIterator on the command stream, and then .insert() and .remove() as needed to get the requested translation. It's fairly generic around MessageBoxes, and a little bit stricter around Choices. I think the comments do a reasonable job of explaining what's going on, but let me know if there are some areas that are not clear and I'll tidy up the docs.

I did a pass to simplify the function (a month ago); it's possible it can be simplified a little further, or we could encapsulate some of these activities (e.g., "trim down to 4 lines") into their own tiny functions, if that helps.